File: apps/mobile/lib/services/confidence/enhanced_confidence_service.dart
/// Mobile confidence: 3 axes (completeness, accuracy, freshness), weighted average.
/// Used for UI display when backend is unavailable (offline fallback).
///
/// NOTE: Three confidence systems coexist (V9-5):
///   1. enhanced_confidence_service.py (backend) — 4-axis geometric mean, authoritative
///   2. This file (mobile) — 3-axis weighted average, offline fallback for UI display
///   3. confidence_scorer.dart (financial_core) — project-level scoring with bloc breakdown
///
/// TODO: Unify to single 4-axis model matching backend (SOT §3)
///
/// Weights: completeness 40% + accuracy 35% + freshness 25% = overall
///
/// References:
/// - DATA_ACQUISITION_STRATEGY.md, "Confidence Scoring Evolution"
/// - ADR-20260223-unified-financial-engine.md
library;

// ────────────────────────────────────────────────────────────
//  DATA SOURCE ENUM
// ────────────────────────────────────────────────────────────

/// How a profile field was acquired — drives accuracy scoring.
enum DataSource {
  /// MINT computed default (lowest confidence).
  systemEstimate,

  /// User typed a value without validation.
  userEntry,

  /// User typed a value that passed cross-validation checks.
  userEntryCrossValidated,

  /// Value extracted from a scanned document (not yet confirmed).
  documentScan,

  /// Value extracted from a scanned document AND user-confirmed.
  documentScanVerified,

  /// Live feed from Open Banking (bLink / SFTI).
  openBanking,

  /// Direct feed from institutional API (caisse de pension, AFC).
  institutionalApi,
}

// ────────────────────────────────────────────────────────────
//  FIELD SOURCE — tracks provenance of a single profile field
// ────────────────────────────────────────────────────────────

/// Metadata for a single profile field: its value, where it came from,
/// and when it was last updated.
class FieldSource {
  final String fieldName;
  final DataSource source;
  final DateTime updatedAt;
  final dynamic value;

  const FieldSource({
    required this.fieldName,
    required this.source,
    required this.updatedAt,
    required this.value,
  });
}

// ────────────────────────────────────────────────────────────
//  CONFIDENCE BREAKDOWN — 3-axis score
// ────────────────────────────────────────────────────────────

/// Three-dimensional confidence score.
///
/// - **completeness** (0-100): how many expected fields are filled
/// - **accuracy** (0-100): weighted quality of data sources
/// - **freshness** (0-100): how recent the data is
/// - **overall**: weighted 40/35/25
class ConfidenceBreakdown {
  final double completeness;
  final double accuracy;
  final double freshness;

  /// Weighted combination: 40% completeness + 35% accuracy + 25% freshness.
  double get overall =>
      (completeness * 0.40 + accuracy * 0.35 + freshness * 0.25)
          .clamp(0.0, 100.0);

  const ConfidenceBreakdown({
    required this.completeness,
    required this.accuracy,
    required this.freshness,
  });
}

// ────────────────────────────────────────────────────────────
//  ENRICHMENT PROMPT — actionable step to improve confidence
// ────────────────────────────────────────────────────────────

/// Suggested user action to improve confidence score.
class EnrichmentPrompt {
  /// Profile field that would be improved.
  final String fieldName;

  /// Human-readable action description (FR).
  final String action;

  /// How many confidence points this action would add.
  final int impactPoints;

  /// Acquisition method (e.g. 'documentScan', 'manualEntry', 'openBanking').
  final String method;

  /// Priority rank (1 = highest).
  final int priority;

  const EnrichmentPrompt({
    required this.fieldName,
    required this.action,
    required this.impactPoints,
    required this.method,
    required this.priority,
  });
}

// ────────────────────────────────────────────────────────────
//  FEATURE GATE — what's unlocked at each confidence level
// ────────────────────────────────────────────────────────────

/// A feature that requires a minimum confidence to unlock.
class FeatureGate {
  final String gateName;
  final bool unlocked;
  final double minConfidence;

  const FeatureGate({
    required this.gateName,
    required this.unlocked,
    required this.minConfidence,
  });
}

// ────────────────────────────────────────────────────────────
//  CONFIDENCE RESULT — complete output
// ────────────────────────────────────────────────────────────

/// Complete enhanced confidence result.
class ConfidenceResult {
  final ConfidenceBreakdown breakdown;
  final List<EnrichmentPrompt> enrichmentPrompts;
  final List<FeatureGate> featureGates;
  final String disclaimer;
  final List<String> sources;

  const ConfidenceResult({
    required this.breakdown,
    required this.enrichmentPrompts,
    required this.featureGates,
    required this.disclaimer,
    required this.sources,
  });
}

// ────────────────────────────────────────────────────────────
//  ENHANCED CONFIDENCE SERVICE — static, pure functions
// ────────────────────────────────────────────────────────────

/// Enhanced confidence scorer with 3 axes: completeness, accuracy, freshness.
///
/// All methods are static and deterministic. This service mirrors the backend
/// enhanced confidence scoring logic.
class EnhancedConfidenceService {
  EnhancedConfidenceService._();

  // ── Accuracy weights per DataSource (backend-aligned) ──────

  static const Map<DataSource, double> _accuracyWeights = {
    DataSource.openBanking: 1.00,
    DataSource.institutionalApi: 0.95,
    DataSource.documentScanVerified: 0.95,
    DataSource.documentScan: 0.85,
    DataSource.userEntryCrossValidated: 0.70,
    DataSource.userEntry: 0.50,
    DataSource.systemEstimate: 0.25,
  };

  // Backend-aligned field importance weights.
  static const Map<String, double> _profileFieldWeights = {
    'age': 1.0,
    'canton': 0.8,
    'salaire_brut': 1.0,
    'salaire_net': 0.6,
    'lpp_total': 1.0,
    'lpp_obligatoire': 1.0,
    'lpp_surobligatoire': 0.8,
    'lpp_insured_salary': 0.7,
    'conversion_rate_oblig': 0.9,
    'conversion_rate_suroblig': 0.7,
    'buyback_potential': 0.6,
    'employee_lpp_contribution': 0.5,
    'avs_contribution_years': 0.9,
    'avs_ramd': 0.9,
    'pillar_3a_balance': 0.7,
    'taux_marginal': 0.9,
    'taxable_income': 0.7,
    'taxable_wealth': 0.5,
    'mortgage_remaining': 0.5,
    'mortgage_rate': 0.4,
    'property_value': 0.4,
    'is_married': 0.5,
    'nb_children': 0.4,
    'monthly_expenses': 0.6,
    'is_independant': 0.6,
    'has_lpp': 0.7,
  };

  // ────────────────────────────────────────────────────────────
  //  1. COMPLETENESS (0-100)
  // ────────────────────────────────────────────────────────────

  /// Scores how many of the expected profile fields are filled.
  ///
  /// Each field has an importance weight; the score is the weighted
  /// sum of filled fields normalized to 0-100.
  static double scoreCompleteness(Map<String, dynamic> profile) {
    final totalWeight =
        _profileFieldWeights.values.fold(0.0, (sum, w) => sum + w);

    double scored = 0;
    for (final entry in _profileFieldWeights.entries) {
      final val = profile[entry.key];
      if (_isFilled(val)) {
        scored += entry.value;
      }
    }

    return ((scored / totalWeight) * 100).clamp(0.0, 100.0);
  }

  // ────────────────────────────────────────────────────────────
  //  2. ACCURACY (0-100)
  // ────────────────────────────────────────────────────────────

  /// Scores the weighted average quality of data sources.
  ///
  /// Fields without a [FieldSource] entry are assumed to be `systemEstimate`.
  static double scoreAccuracy(List<FieldSource> fieldSources) {
    if (fieldSources.isEmpty) return 0;

    double weightedSum = 0;
    double totalWeight = 0;

    for (final fs in fieldSources) {
      final sourceAccuracy = _accuracyWeights[fs.source] ?? 0.25;
      final fieldWeight = _profileFieldWeights[fs.fieldName] ?? 0.5;
      weightedSum += sourceAccuracy * fieldWeight;
      totalWeight += fieldWeight;
    }

    if (totalWeight == 0) return 0;
    return ((weightedSum / totalWeight) * 100).clamp(0.0, 100.0);
  }

  // ────────────────────────────────────────────────────────────
  //  3. FRESHNESS (0-100)
  // ────────────────────────────────────────────────────────────

  /// Scores the average freshness of field data.
  ///
  /// Freshness decay:
  /// - < 1 month: 1.00
  /// - 1-3 months: 0.90
  /// - 3-6 months: 0.75
  /// - 6-12 months: 0.50
  /// - > 12 months: 0.25
  static double scoreFreshness(List<FieldSource> fieldSources) {
    if (fieldSources.isEmpty) return 0;

    final now = DateTime.now();
    double weightedFreshness = 0;
    double totalWeight = 0;

    for (final fs in fieldSources) {
      final age = now.difference(fs.updatedAt);
      final fieldWeight = _profileFieldWeights[fs.fieldName] ?? 0.5;
      weightedFreshness += _freshnessForAge(age) * fieldWeight;
      totalWeight += fieldWeight;
    }

    if (totalWeight == 0) return 0;
    return ((weightedFreshness / totalWeight) * 100).clamp(0.0, 100.0);
  }

  /// Returns freshness score (0.0 - 1.0) for a given data age.
  static double _freshnessForAge(Duration age) {
    final days = age.inDays;
    if (days < 30) return 1.00;
    if (days < 90) return 0.90;
    if (days < 180) return 0.75;
    if (days < 365) return 0.50;
    return 0.25;
  }

  // ────────────────────────────────────────────────────────────
  //  4. COMPUTE CONFIDENCE (full result)
  // ────────────────────────────────────────────────────────────

  /// Computes the full enhanced confidence result.
  ///
  /// [profile] — key-value map of the user's profile fields.
  /// [fieldSources] — provenance metadata per field.
  static ConfidenceResult computeConfidence(
    Map<String, dynamic> profile,
    List<FieldSource> fieldSources,
  ) {
    final completeness = scoreCompleteness(profile);
    final accuracy = scoreAccuracy(fieldSources);
    final freshness = scoreFreshness(fieldSources);

    final breakdown = ConfidenceBreakdown(
      completeness: completeness,
      accuracy: accuracy,
      freshness: freshness,
    );

    final prompts = rankEnrichmentPrompts(profile, fieldSources);
    final gates = _computeFeatureGates(breakdown.overall);

    return ConfidenceResult(
      breakdown: breakdown,
      enrichmentPrompts: prompts,
      featureGates: gates,
      disclaimer:
          'Outil educatif — ne constitue pas un conseil financier (LSFin). '
          'Le score de confiance reflete la qualite des donnees fournies, '
          'pas la fiabilite du systeme.',
      sources: const [
        'DATA_ACQUISITION_STRATEGY.md',
        'LPP art. 7-16',
        'LAVS art. 21-40',
        'LIFD art. 38',
        'OPP3 art. 7',
      ],
    );
  }

  // ────────────────────────────────────────────────────────────
  //  5. RANK ENRICHMENT PROMPTS
  // ────────────────────────────────────────────────────────────

  /// Returns enrichment prompts ranked by impact on overall confidence.
  ///
  /// Prompts are generated for missing or low-quality fields and sorted
  /// by descending impact.
  static List<EnrichmentPrompt> rankEnrichmentPrompts(
    Map<String, dynamic> profile,
    List<FieldSource> fieldSources,
  ) {
    final prompts = <EnrichmentPrompt>[];
    final sourceMap = <String, FieldSource>{};
    for (final fs in fieldSources) {
      sourceMap[fs.fieldName] = fs;
    }

    int priority = 1;

    // --- LPP certificate (highest impact) ---
    if (!_isFilled(profile['lpp_obligatoire'])) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'lpp_obligatoire',
        action: 'Scanne ton certificat de prevoyance LPP',
        impactPoints: 27,
        method: 'documentScan',
        priority: priority++,
      ));
    } else {
      final source = sourceMap['lpp_obligatoire'];
      if (source != null && _isLowQuality(source)) {
        prompts.add(EnrichmentPrompt(
          fieldName: 'lpp_obligatoire',
          action: 'Confirme ta part obligatoire LPP avec ton certificat',
          impactPoints: 15,
          method: 'documentScanVerified',
          priority: priority++,
        ));
      }
    }

    // --- LPP total ---
    if (!_isFilled(profile['lpp_total'])) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'lpp_total',
        action: 'Ajoute ton avoir LPP total',
        impactPoints: 20,
        method: 'documentScan',
        priority: priority++,
      ));
    }

    // --- AVS extract ---
    if (!_isFilled(profile['avs_contribution_years'])) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'avs_contribution_years',
        action: 'Commande ton extrait AVS (gratuit sur inforegister.ch)',
        impactPoints: 18,
        method: 'documentScan',
        priority: priority++,
      ));
    }

    // --- Taux marginal (tax declaration) ---
    if (!_isFilled(profile['taux_marginal'])) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'taux_marginal',
        action: 'Scanne ton avis de taxation pour le taux marginal',
        impactPoints: 15,
        method: 'documentScan',
        priority: priority++,
      ));
    } else {
      final source = sourceMap['taux_marginal'];
      if (source != null && _isLowQuality(source)) {
        prompts.add(EnrichmentPrompt(
          fieldName: 'taux_marginal',
          action: 'Verifie ton taux marginal avec ton avis de taxation',
          impactPoints: 10,
          method: 'documentScanVerified',
          priority: priority++,
        ));
      }
    }

    // --- Pillar 3a ---
    if (!_isFilled(profile['pillar_3a_balance'])) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'pillar_3a_balance',
        action: 'Renseigne tes soldes 3a (attestation ou e-banking)',
        impactPoints: 10,
        method: 'manualEntry',
        priority: priority++,
      ));
    }

    // --- Patrimoine ---
    if (!_isFilled(profile['patrimoine_total'])) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'patrimoine_total',
        action: 'Ajoute ton patrimoine (epargne, investissements)',
        impactPoints: 8,
        method: 'manualEntry',
        priority: priority++,
      ));
    }

    // --- Salary ---
    if (!_isFilled(profile['salaire_brut'])) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'salaire_brut',
        action: 'Renseigne ton salaire brut mensuel',
        impactPoints: 12,
        method: 'manualEntry',
        priority: priority++,
      ));
    }

    // --- Open Banking opportunity ---
    final hasOb = fieldSources.any((fs) => fs.source == DataSource.openBanking);
    if (!hasOb) {
      prompts.add(EnrichmentPrompt(
        fieldName: 'open_banking',
        action: 'Connecte ton compte bancaire (bLink)',
        impactPoints: 22,
        method: 'openBanking',
        priority: priority++,
      ));
    }

    // Sort by impact descending
    prompts.sort((a, b) => b.impactPoints.compareTo(a.impactPoints));

    // Re-assign priority after sort
    for (int i = 0; i < prompts.length; i++) {
      prompts[i] = EnrichmentPrompt(
        fieldName: prompts[i].fieldName,
        action: prompts[i].action,
        impactPoints: prompts[i].impactPoints,
        method: prompts[i].method,
        priority: i + 1,
      );
    }

    return prompts;
  }

  // ────────────────────────────────────────────────────────────
  //  FEATURE GATES
  // ────────────────────────────────────────────────────────────

  /// Computes which features are unlocked at the current confidence level.
  ///
  /// Gates (from DATA_ACQUISITION_STRATEGY.md):
  /// - < 30%: basic only
  /// - 30-50%: standard projections
  /// - 50-70%: arbitrage with uncertainty bands
  /// - 70-85%: precise arbitrage + FRI
  /// - > 85%: full precision + longitudinal
  static List<FeatureGate> _computeFeatureGates(double overall) {
    return [
      FeatureGate(
        gateName: 'Chiffre choc de base',
        unlocked: overall >= 0,
        minConfidence: 0,
      ),
      FeatureGate(
        gateName: 'Projections standard (3 scenarios)',
        unlocked: overall >= 30,
        minConfidence: 30,
      ),
      FeatureGate(
        gateName: 'Arbitrage avec bandes d\'incertitude',
        unlocked: overall >= 50,
        minConfidence: 50,
      ),
      FeatureGate(
        gateName: 'Arbitrage precis + score FRI',
        unlocked: overall >= 70,
        minConfidence: 70,
      ),
      FeatureGate(
        gateName: 'Suivi longitudinal',
        unlocked: overall >= 85,
        minConfidence: 85,
      ),
    ];
  }

  // ────────────────────────────────────────────────────────────
  //  PRIVATE HELPERS
  // ────────────────────────────────────────────────────────────

  /// Returns true if a profile field value is considered "filled".
  static bool _isFilled(dynamic val) {
    if (val == null) return false;
    if (val is num) {
      if (val is double && val.isNaN) return false;
      return true;
    }
    if (val is String) return val.isNotEmpty;
    return true;
  }

  /// Returns true if a field source is low quality (estimate or raw entry).
  static bool _isLowQuality(FieldSource fs) {
    return fs.source == DataSource.systemEstimate ||
        fs.source == DataSource.userEntry;
  }
}

File: services/backend/app/services/confidence/enhanced_confidence_service.py
"""
# ════════════════════════════════════════════════════════════════════════════
# CONFIDENCE DOCTRINE (see docs/SOURCE_OF_TRUTH_MATRIX.md §3)
#
# This service is the AUTHORITATIVE SOURCE OF TRUTH for confidence scoring.
# It governs: feature gates, global UI confidence bars, enrichment ranking.
#
# The 3 confidence systems and their governance:
#   1. THIS FILE → authoritative, 4-axis geometric mean, via /confidence API
#   2. enhanced_confidence_service.dart (mobile) → offline fallback (3-axis)
#   3. confidence_scorer.dart (financial_core) → projection quality only
#
# Any new confidence-related feature should consume this service via API.
# The mobile fallback exists only for offline UX, not for decision-making.
# ════════════════════════════════════════════════════════════════════════════

Backend confidence: 4 axes (completeness, accuracy, freshness, understanding), geometric mean.
Used by /confidence API endpoint.

Pure functions for multi-dimensional confidence measurement:
- score_completeness: champs remplis, ponderes par importance
- score_accuracy: qualite des sources (open_banking > document > user_entry)
- score_freshness: fraicheur des donnees (< 1 mois = 100, > 12 mois = 25)
- score_understanding: comprehension financiere (literacy + engagement)
- compute_confidence: moyenne geometrique 4 axes + feature gates + enrichment prompts
- rank_enrichment_prompts: actions classees par impact sur le score

Sources:
    - DATA_ACQUISITION_STRATEGY.md, section "Confidence Scoring Evolution"
    - LPP art. 7-8, 14-16 (prevoyance professionnelle)
    - LAVS art. 21-40 (AVS)
    - OPP3 art. 7 (3e pilier)
    - LIFD art. 38 (imposition du capital)
"""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Dict, List, Optional

from app.services.document_parser.document_models import DATA_SOURCE_ACCURACY

from app.services.confidence.enhanced_confidence_models import (
    ConfidenceBreakdown,
    ConfidenceResult,
    EnrichmentPrompt,
    FieldSource,
)


# ============================================================================
# Constants
# ============================================================================

# Weight of each axis in the overall confidence score
WEIGHT_COMPLETENESS = 0.40
WEIGHT_ACCURACY = 0.35
WEIGHT_FRESHNESS = 0.25

# Profile fields with importance weights (higher = more impact on projections)
# Fields critical for retirement, tax, and arbitrage projections are weighted higher.
PROFILE_FIELD_WEIGHTS: Dict[str, float] = {
    # Identity / core
    "age": 1.0,
    "canton": 0.8,
    "salaire_brut": 1.0,
    "salaire_net": 0.6,
    # LPP (2nd pillar) — critical for retirement + arbitrage
    "lpp_total": 1.0,
    "lpp_obligatoire": 1.0,
    "lpp_surobligatoire": 0.8,
    "lpp_insured_salary": 0.7,
    "conversion_rate_oblig": 0.9,
    "conversion_rate_suroblig": 0.7,
    "buyback_potential": 0.6,
    "employee_lpp_contribution": 0.5,
    # AVS (1st pillar)
    "avs_contribution_years": 0.9,
    "avs_ramd": 0.9,
    # 3a (pillar 3a)
    "pillar_3a_balance": 0.7,
    # Tax
    "taux_marginal": 0.9,
    "taxable_income": 0.7,
    "taxable_wealth": 0.5,
    # Mortgage / property
    "mortgage_remaining": 0.5,
    "mortgage_rate": 0.4,
    "property_value": 0.4,
    # Family / situation
    "is_married": 0.5,
    "nb_children": 0.4,
    "monthly_expenses": 0.6,
    # Employment
    "is_independant": 0.6,
    "has_lpp": 0.7,
}

# Freshness decay thresholds (months -> score)
FRESHNESS_THRESHOLDS = [
    (1, 1.00),    # < 1 month
    (3, 0.90),    # 1-3 months
    (6, 0.75),    # 3-6 months
    (12, 0.50),   # 6-12 months
]
FRESHNESS_FLOOR = 0.25  # > 12 months

# Understanding axis constants (Dart parity)
LITERACY_BASES = {"beginner": 30.0, "intermediate": 55.0, "advanced": 85.0}
SESSION_BONUS_PER_CHECKIN = 2.0
SESSION_BONUS_CAP = 40.0
WEIGHT_LITERACY = 0.50
WEIGHT_SESSION = 0.30
WEIGHT_EDUCATION = 0.20

# Enrichment prompt catalog: what actions improve confidence and by how much
_ENRICHMENT_CATALOG: List[Dict] = [
    {
        "field_name": "lpp_total",
        "action": "Scanne ton certificat de prevoyance LPP",
        "impact_points": 27.0,
        "method": "document_scan",
        "related_fields": [
            "lpp_obligatoire", "lpp_surobligatoire", "lpp_insured_salary",
            "conversion_rate_oblig", "conversion_rate_suroblig",
            "buyback_potential", "employee_lpp_contribution",
        ],
    },
    {
        "field_name": "avs_contribution_years",
        "action": "Demande ton extrait de compte individuel AVS (ahv-iv.ch)",
        "impact_points": 22.0,
        "method": "avs_request",
        "related_fields": ["avs_ramd"],
    },
    {
        "field_name": "taux_marginal",
        "action": "Scanne ta declaration fiscale ou ton avis de taxation",
        "impact_points": 18.0,
        "method": "document_scan",
        "related_fields": ["taxable_income", "taxable_wealth"],
    },
    {
        "field_name": "salaire_brut",
        "action": "Connecte ton compte bancaire via Open Banking",
        "impact_points": 20.0,
        "method": "open_banking",
        "related_fields": [
            "salaire_net", "monthly_expenses", "pillar_3a_balance",
        ],
    },
    {
        "field_name": "pillar_3a_balance",
        "action": "Scanne ton attestation 3a ou connecte ton compte 3a",
        "impact_points": 8.0,
        "method": "document_scan",
        "related_fields": [],
    },
    {
        "field_name": "mortgage_remaining",
        "action": "Scanne ton attestation hypothecaire",
        "impact_points": 12.0,
        "method": "document_scan",
        "related_fields": ["mortgage_rate", "property_value"],
    },
    {
        "field_name": "monthly_expenses",
        "action": "Entre tes charges mensuelles ou connecte Open Banking",
        "impact_points": 6.0,
        "method": "manual_entry",
        "related_fields": [],
    },
    {
        "field_name": "is_married",
        "action": "Complete ta situation familiale (marie-e, enfants)",
        "impact_points": 5.0,
        "method": "manual_entry",
        "related_fields": ["nb_children"],
    },
]


# ============================================================================
# Pure functions
# ============================================================================


def score_completeness(profile: dict) -> float:
    """Calcule le score de completude (0-100) du profil.

    Chaque champ est pondere par son importance pour les projections financieres.
    Un profil completement rempli = 100. Un profil vide = 0.

    Args:
        profile: Dictionnaire du profil utilisateur (cles = noms de champs).

    Returns:
        Score de completude 0-100.
    """
    if not PROFILE_FIELD_WEIGHTS:
        return 0.0

    total_weight = sum(PROFILE_FIELD_WEIGHTS.values())
    filled_weight = 0.0

    for field_name, weight in PROFILE_FIELD_WEIGHTS.items():
        if _is_field_filled(profile.get(field_name)):
            filled_weight += weight

    if total_weight == 0:
        return 0.0

    return round(min(100.0, (filled_weight / total_weight) * 100), 1)


def _is_field_filled(value: object) -> bool:
    """Retourne True si une valeur de profil est consideree comme renseignee.

    Regles:
    - bool: True et False sont tous deux renseignes
    - None / "": non renseignes
    - nombres (incl. 0): renseignes
    - autres types: truthy => renseignes
    """
    if isinstance(value, bool):
        return True
    if value is None:
        return False
    if isinstance(value, str):
        return value.strip() != ""
    if isinstance(value, (int, float)):
        return True
    return bool(value)


def score_accuracy(field_sources: List[FieldSource]) -> float:
    """Calcule le score de precision (0-100) base sur la qualite des sources.

    Chaque source a un poids de fiabilite (open_banking=1.0, user_entry=0.5, etc.).
    Le score est la moyenne ponderee des accuracy weights, ponderee par l'importance
    du champ dans le profil.

    Args:
        field_sources: Liste des sources de champs du profil.

    Returns:
        Score de precision 0-100.
    """
    if not field_sources:
        return 0.0

    weighted_accuracy_sum = 0.0
    total_weight = 0.0

    for fs in field_sources:
        field_weight = PROFILE_FIELD_WEIGHTS.get(fs.field_name, 0.5)
        source_accuracy = DATA_SOURCE_ACCURACY.get(fs.source, 0.25)
        weighted_accuracy_sum += source_accuracy * field_weight
        total_weight += field_weight

    if total_weight == 0:
        return 0.0

    return round(min(100.0, (weighted_accuracy_sum / total_weight) * 100), 1)


def _compute_field_freshness(updated_at: str, now: Optional[datetime] = None) -> float:
    """Calcule la fraicheur d'un champ individuel (0-1).

    Decay:
        < 1 mois:   1.00
        1-3 mois:   0.90
        3-6 mois:   0.75
        6-12 mois:  0.50
        > 12 mois:  0.25

    Args:
        updated_at: Date ISO 8601 de derniere mise a jour.
        now: Date de reference (default: maintenant UTC).

    Returns:
        Score de fraicheur 0-1.
    """
    if now is None:
        now = datetime.now(timezone.utc)

    try:
        # Parse ISO datetime — handle both with and without timezone
        dt = datetime.fromisoformat(updated_at.replace("Z", "+00:00"))
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
    except (ValueError, AttributeError):
        return FRESHNESS_FLOOR

    age_days = (now - dt).total_seconds() / 86400.0
    if age_days < 0:
        # Future date — treat as fresh
        return 1.0

    age_months = age_days / 30.44  # average days per month

    for threshold_months, score in FRESHNESS_THRESHOLDS:
        if age_months < threshold_months:
            return score

    return FRESHNESS_FLOOR


def score_freshness(
    field_sources: List[FieldSource],
    now: Optional[datetime] = None,
) -> float:
    """Calcule le score de fraicheur (0-100) des donnees du profil.

    Moyenne ponderee des freshness scores par champ, ponderee par importance.

    Args:
        field_sources: Liste des sources de champs du profil.
        now: Date de reference (default: maintenant UTC).

    Returns:
        Score de fraicheur 0-100.
    """
    if not field_sources:
        return 0.0

    weighted_freshness_sum = 0.0
    total_weight = 0.0

    for fs in field_sources:
        field_weight = PROFILE_FIELD_WEIGHTS.get(fs.field_name, 0.5)
        field_fresh = _compute_field_freshness(fs.updated_at, now=now)
        weighted_freshness_sum += field_fresh * field_weight
        total_weight += field_weight

    if total_weight == 0:
        return 0.0

    return round(min(100.0, (weighted_freshness_sum / total_weight) * 100), 1)


def score_understanding(profile: dict) -> float:
    """Compute understanding score (0-100) from literacy level and engagement.

    Formula: literacyBase * 0.50 + sessionBonus * 0.30 + educationBonus * 0.20

    Args:
        profile: dict with optional keys 'financial_literacy_level' and 'check_in_count'

    Returns:
        Understanding score 0-100.
    """
    literacy_level = profile.get("financial_literacy_level", "beginner")
    literacy_base = LITERACY_BASES.get(str(literacy_level).lower(), 30.0)

    check_in_count = profile.get("check_in_count", 0)
    if not isinstance(check_in_count, (int, float)):
        check_in_count = 0
    session_bonus = min(check_in_count * SESSION_BONUS_PER_CHECKIN, SESSION_BONUS_CAP)

    education_bonus = 0.0  # placeholder for future education tracking

    understanding = (
        literacy_base * WEIGHT_LITERACY
        + session_bonus * WEIGHT_SESSION
        + education_bonus * WEIGHT_EDUCATION
    )
    return round(min(100.0, max(0.0, understanding)), 1)


def _geo_mean_4(c: float, a: float, f: float, u: float) -> float:
    """4-axis geometric mean with shift to avoid zero (Dart parity)."""
    vals = [(x + 1.0) / 101.0 for x in [c, a, f, u]]
    product = vals[0] * vals[1] * vals[2] * vals[3]
    geo = product ** 0.25
    return min(100.0, max(0.0, geo * 101.0 - 1.0))


def _compute_feature_gates(overall: float) -> Dict[str, bool]:
    """Determine les fonctionnalites debloquees selon le score global.

    Feature gates:
        < 30%:  basic_premier_eclairage_only
        30-50%: + standard_projections
        50-70%: + arbitrage_comparisons (with uncertainty bands)
        70-85%: + precise_arbitrage + fri_scoring
        > 85%:  + full_precision + longitudinal_tracking

    Args:
        overall: Score de confiance global (0-100).

    Returns:
        Dict de feature gates (nom -> actif/inactif).
    """
    return {
        "basic_premier_eclairage_only": True,  # always available
        "standard_projections": overall >= 30.0,
        "arbitrage_comparisons": overall >= 50.0,
        "precise_arbitrage": overall >= 70.0,
        "fri_scoring": overall >= 70.0,
        "full_precision": overall >= 85.0,
        "longitudinal_tracking": overall >= 85.0,
    }


def rank_enrichment_prompts(
    profile: dict,
    field_sources: List[FieldSource],
) -> List[EnrichmentPrompt]:
    """Classe les actions d'enrichissement par impact decroissant.

    Pour chaque action du catalogue, calcule l'impact reel en fonction
    de ce qui est deja rempli/source dans le profil.

    Actions deja completees (champ rempli + source de bonne qualite) sont
    exclues ou ont un impact reduit.

    Args:
        profile: Dictionnaire du profil utilisateur.
        field_sources: Liste des sources existantes.

    Returns:
        Liste d'EnrichmentPrompt classee par impact decroissant.
    """
    # Build lookup for existing sources
    source_by_field: Dict[str, FieldSource] = {
        fs.field_name: fs for fs in field_sources
    }

    prompts: List[EnrichmentPrompt] = []

    for catalog_entry in _ENRICHMENT_CATALOG:
        primary_field = catalog_entry["field_name"]
        base_impact = catalog_entry["impact_points"]
        related_fields = catalog_entry["related_fields"]

        # Check if primary field is already well-sourced
        existing_source = source_by_field.get(primary_field)
        primary_filled = profile.get(primary_field) is not None and profile.get(primary_field) != ""

        if primary_filled and existing_source:
            # Field already has a good source — reduce impact
            source_quality = DATA_SOURCE_ACCURACY.get(existing_source.source, 0.25)
            if source_quality >= 0.85:
                # Already well sourced (document_scan or better) — skip
                continue
            # Partial credit: reduce impact proportionally
            base_impact *= (1.0 - source_quality)

        # Check related fields: if many are missing, higher impact
        missing_related = sum(
            1 for rf in related_fields
            if profile.get(rf) is None or profile.get(rf) == ""
        )
        total_related = len(related_fields) if related_fields else 1

        # Bonus for filling multiple missing related fields
        related_bonus = (missing_related / max(total_related, 1)) * base_impact * 0.3

        effective_impact = round(base_impact + related_bonus, 1)

        if effective_impact < 1.0:
            continue

        prompts.append(EnrichmentPrompt(
            field_name=primary_field,
            action=catalog_entry["action"],
            impact_points=effective_impact,
            method=catalog_entry["method"],
            priority=0,  # will be set below
        ))

    # Sort by impact descending
    prompts.sort(key=lambda p: p.impact_points, reverse=True)

    # Assign priority ranks (1 = highest impact)
    for i, prompt in enumerate(prompts):
        prompt.priority = i + 1

    return prompts


def compute_confidence(
    profile: dict,
    field_sources: List[FieldSource],
    now: Optional[datetime] = None,
) -> ConfidenceResult:
    """Calcule le score de confiance complet sur 4 axes.

    Combine completeness, accuracy, freshness et understanding
    via une moyenne geometrique, genere les feature gates et classe les
    enrichment prompts par impact.

    Args:
        profile: Dictionnaire du profil utilisateur.
        field_sources: Liste des sources de champs.
        now: Date de reference pour la fraicheur (default: maintenant UTC).

    Returns:
        ConfidenceResult complet avec breakdown, gates, prompts et compliance.
    """
    completeness = score_completeness(profile)
    accuracy = score_accuracy(field_sources)
    freshness = score_freshness(field_sources, now=now)
    understanding = score_understanding(profile)

    # 4-axis geometric mean (Dart parity)
    overall = round(_geo_mean_4(completeness, accuracy, freshness, understanding), 1)

    breakdown = ConfidenceBreakdown(
        completeness=completeness,
        accuracy=accuracy,
        freshness=freshness,
        understanding=understanding,
        overall=overall,
    )

    feature_gates = _compute_feature_gates(overall)
    enrichment_prompts = rank_enrichment_prompts(profile, field_sources)

    return ConfidenceResult(
        breakdown=breakdown,
        enrichment_prompts=enrichment_prompts,
        feature_gates=feature_gates,
    )
